compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -4407,6 +4408,30 @@ declare_lint! {
     "unrecognized diagnostic attribute"
 }
 
+declare_lint! {
+    /// The `malformed_diagnostic_attributes` lint detects malformed diagnostic attributes.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// #![feature(diagnostic_namespace)]
+    /// #[diagnostic::on_unimplemented]
+    /// struct Foo;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to specify invalid options to an existing diagnostic attribute.
+    /// Check the spelling, and check the diagnostic attribute listing for the correct set of options.
+    /// Also consider if you are using an old version of the compiler. Certain options may only
+    /// be available on newer compiler versions
+    pub MALFORMED_DIAGNOSTIC_ATTRIBUTES,
+    Warn,
+    "unrecognized diagnostic attribute options"
```

Comment:
There are too many new lints for this feature, IMO.

Is it ever necessary to e.g. allow malformed diagnostic attributes, but warn on unknown?
I don't think so, this lints generally exists to warn on diagnostic attributes that are both known and well-formed, just on some other version of Rust compiler, so the distinction is not significant.

Reply:
Also the previous lint was introduced as stable, but it should probably use a feature gate `@feature_gate = sym::feature_gate_name;` for now.

Reply:
:+1: Will change that to use the other lint. We might consider renaming the lint then.

For the `@feature_gate` change: Is it ok to put that into this PR as well or should that be opened as a separate PR?

Reply:
I've pushed https://github.com/rust-lang/rust/pull/114452/commits/c59ef01eee6572063e357560d619f51d53c73005 to address these concerns.